edgedb/asyncio_con.py

```diff
@@ -32,41 +35,160 @@
 from .datatypes import datatypes
 from .protocol import asyncio_proto
 from .protocol import protocol
+from .protocol.protocol import CodecsRegistry as _CodecsRegistry
+from .protocol.protocol import QueryCodecsCache as _QueryCodecsCache
+
+
+ERRNO_RE = re.compile(r"\[Errno (\d+)\]")
+TEMPORARY_ERRORS = frozenset({
+    errno.ECONNREFUSED,
+    errno.ECONNABORTED,
+    errno.ECONNRESET,
+    errno.ENOENT,
+})
+
+
+def _extract_errno(s):
+    """Extract multiple errnos from error string
+
+    When we connect to a host that has multiple underlying IP addresses, say
+    ``localhost`` having ``::1`` and ``127.0.0.1``, we get
+    ``OSError("Multiple exceptions:...")`` error without ``.errno`` attribute
+    set. There are multiple ones in the text, so we extract all of them.
+    """
+    result = []
+    for match in ERRNO_RE.finditer(s):
+        result.append(int(match.group(1)))
+    if result:
+        return result
 
 
-class _ConnectionProxy:
-    # Base class to enable `isinstance(AsyncIOConnection)` check.
-    __slots__ = ()
+class _AsyncIOConnectionImpl:
 
+    def __init__(self):
+        self._addr = None
+        self._transport = None
+        self._protocol = None
 
-class AsyncIOConnectionMeta(abc.ABCMeta):
+    def is_closed(self):
+        transport = self._transport
+        protocol = self._protocol
+        return not (
+            transport and not transport.is_closing() and
+            protocol and protocol.connected
+        )
 
-    def __instancecheck__(cls, instance):
-        mro = type(instance).__mro__
-        return AsyncIOConnection in mro or _ConnectionProxy in mro
+    async def connect(self, loop, addrs, config, params):
+        addr = None
+        max_time = time.monotonic() + config.wait_until_available
+        iteration = 1
+
+        while True:
+            for addr in addrs:
+                try:
+                    await asyncio.wait_for(
+                        self._connect_addr(loop, addr, params),
+                        config.connect_timeout,
+                    )
+                except asyncio.TimeoutError as e:
+                    if iteration == 1 or time.monotonic() < max_time:
```

Comment:
Re `iteration == 1 or ...`: so if `wait_until_available` is 0 we'd still give it two tries?

Reply:
Iterations start from `1`. When `wait_until_available` is zero we give a one try for each of the addresses specified.